### DISCUSSION TITLE
MyPy

### DISCUSSION THREAD

**Discussion by cbusillo** [trinhanhngoc/odoo-stubs#19] [General]
Does anyone use Odoo-stubs with MyPy?  I am finding that PyCharm just isn't strict enough.  I love the way the PyCharm Odoo plugin works, but I can't seem to get mypy to work well with Odoo.  Is there a good mypy.ini that will help?  I have Odoo-stubs installed with pip install -e ., but I still get lots of issues, for example

```python
    _name = "shopify.client"
    _description = "Shopify GraphQL API Client"
```

```
Mypy: Incompatible types in assignment (expression has type "str", base class "BaseModel" defined the type as "None")  [assignment]
Mypy: Name "odoo" is not defined  [name-defined]

```

**Comment by trinhanhngoc** [comment]
Hello @cbusillo ,

I think currently we cannot get mypy to work well with Odoo without developing a mypy plugin for Odoo. The situation may get better in the future as Odoo is improved to work well with popular static type checkers for Python.

**Comment by cbusillo** [comment]
Too bad.  Mypy is so helpful for things like missing returns and missing type hints on parameters etc. Are there any other static type checkers you have found to help with Odoo?

BTW, your plugin is like magic.  I truly love it.  I'm not sure I would have figured out the nuances of Odoo without it.

**Reply by trinhanhngoc** [reply]
Mypy and Pyright are the two best static type checkers I know. Currently, neither of them work well with Odoo. I think we need to wait for Odoo to improve itself (https://github.com/odoo/odoo/pull/103227, https://github.com/odoo/odoo/pull/167307, https://github.com/odoo/odoo/pull/173996,...).

**Comment by cbusillo** [comment]
Thank you for the information, keep up the good work.
